[#9564] Consolidate GET feedbacksessions RESTFul API

junming403 opened the pull request:
Part of #9564 

Outline of solution:  Get list of feedback sessions according to argument
```
Instructor:
	1. with course id: return all sessions associated with the course 
	2. without course id: 
		2.1 recyle bin flag = true, return all sessions associated with the account in the recycle bin
		2.2 recyle bin flag = false, return all sessions associated with the account except in the recycle bin.

Student:
	1. with course id: return all sessions associated with the course
	2. without course id: return all sessions associated with the account
```

ChooJeremy commented on src/main/java/teammates/ui/webapi/action/GetFeedbackSessionsAction.java:
```diff
@@ -20,12 +23,27 @@ protected AuthType getMinAuthLevel() {
-            gateKeeper.verifyAccessible(logic.getInstructorForGoogleId(courseId, userInfo.getId()), courseAttributes);
+        if (userInfo.isInstructor) {
+            if (courseId == null) {
+                gateKeeper.verifyInstructorPrivileges(logic.getAccount(userInfo.getId()));
```
Is this really necessary? Doesn't usesrInfo.isInstructor perform this check already?

ChooJeremy commented on src/main/java/teammates/ui/webapi/action/GetFeedbackSessionsAction.java:
```diff
@@ -20,12 +23,27 @@ protected AuthType getMinAuthLevel() {
+
+        if (userInfo.isStudent) {
+            if (courseId == null) {
+                gateKeeper.verifyStudentPrivileges(logic.getAccount(userInfo.getId()));
```
Same here, userInfo.isStudent verifies this?

ChooJeremy commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +24,126 @@ protected String getRequestMethod() {
+        GetFeedbackSessionsAction a = getAction(submissionParam);
+        FeedbackSessionsData fsData = (FeedbackSessionsData) getJsonResult(a).getOutput();
+
+        assertEquals(6, fsData.getFeedbackSessions().size());
```
You should verify that there are 6 feedback sessions associated with the course. Ideally, maybe also verify that the feedback sessions actually match?

ChooJeremy commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +24,126 @@ protected String getRequestMethod() {
+        GetFeedbackSessionsAction a = getAction(submissionParam);
+        FeedbackSessionsData fsData = (FeedbackSessionsData) getJsonResult(a).getOutput();
+
+        assertEquals(0, fsData.getFeedbackSessions().size());
```
* Verify that there are actually 0 sessions?
* Have another test for when there is >0 soft-deleted sessions?

ChooJeremy commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +24,126 @@ protected String getRequestMethod() {
+    }
+
+    @Test
+    protected void testExecute_asInstructorWithRecycleBinFlagTrue_shouldReturnAllSoftDeletedSessionsForCourse() {
```
wrong name?
```suggestion
    protected void testExecute_asInstructorWithRecycleBinFlagTrue_shouldReturnAllSoftDeletedSessionsForInstructor() {
```

ChooJeremy commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +24,126 @@ protected String getRequestMethod() {
+    }
+
+    @Test
+    protected void testExecute_asInstructorWithRecycleBinFlagFalse_shouldReturnAllSoftDeletedSessionsForCourse() {
```
wrong name? 
```suggestion
    protected void testExecute_asInstructorWithRecycleBinFlagFalse_shouldReturnSessionsForInstructor() {
```

ChooJeremy commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +24,126 @@ protected String getRequestMethod() {
+        GetFeedbackSessionsAction a = getAction(submissionParam);
+        FeedbackSessionsData fsData = (FeedbackSessionsData) getJsonResult(a).getOutput();
+
+        assertEquals(6, fsData.getFeedbackSessions().size());
```
You should:
* Verify that there are actually 6 non-deleted feedback sessions
* Ideally, pick someone with >0 soft-deleted feedback sessions
* Ideally, verify that feedback sessions match.

ChooJeremy commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +24,126 @@ protected String getRequestMethod() {
+        GetFeedbackSessionsAction a = getAction(submissionParam);
+        FeedbackSessionsData fsData = (FeedbackSessionsData) getJsonResult(a).getOutput();
+
+        assertEquals(6, fsData.getFeedbackSessions().size());
```
Same verification here as per previous comments

ChooJeremy commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +24,126 @@ protected String getRequestMethod() {
+        GetFeedbackSessionsAction a = getAction();
+        FeedbackSessionsData fsData = (FeedbackSessionsData) getJsonResult(a).getOutput();
+
+        assertEquals(6, fsData.getFeedbackSessions().size());
```
Same verification here as per previous comments

junming403 commented:
Updated according to review, the access control logic has been simplified. 
The content of feedback sessions list has also been verified now.

xpdavid commented on src/main/java/teammates/ui/webapi/action/GetFeedbackSessionsAction.java:
```diff
@@ -35,14 +45,22 @@ public ActionResult execute() {
         List<FeedbackSessionAttributes> feedbackSessionAttributes;
         if (courseId == null) {
-            boolean isInRecycleBin = getBooleanRequestParamValue(Const.ParamsNames.IS_IN_RECYCLE_BIN);
+            if (userInfo.isInstructor) {
```
If a login entity is both a student and an instructor of the same course, he will never be able to get courses involved as a student.

Maybe we want to use the presence of the param `IS_IN_RECYCLE_BIN ` to distinguish the request.

junming403 commented:
Updated according to review, for now request specifying `IS_IN_RECYCLE_BIN` are regarded as requests as instructor, absence of this param means it is request as student. For an account be both instructor and student, in order to make request as student, the one needs to have `IS_IN_RECYCLE_BIN` present no matter you specify course_id or not. Updated tests accordingly. 

Ready for review.

xpdavid commented on src/main/java/teammates/ui/webapi/action/GetFeedbackSessionsAction.java:
```diff
@@ -20,12 +23,25 @@ protected AuthType getMinAuthLevel() {
 
     @Override
     public void checkSpecificAccessControl() {
```
Actually, you don't need have too many if here.
```java
if (IS_IN_RECYCLE_BIN == null) {
    if (courseId = null) {
        assert userInfo.isStudent
    } else {
        assert student can access course
    }
} else {
    if (courseId = null) {
        assert userInfo.isInstructor
    } else {
        assert instructor can access course
    }
}
```

Having `userInfo.isInstructor && ! isRequestAsStudent` is confusing especially for new contributor (if you don't explain the situation in the comment)

junming403 replied:
I have looked at the logic again and finds that it indeed can be simplified.  Especially `userInfo.isInstructor` is no longer necessary to check in action since it already passed accessControl.

junming403 commented:
Ready for review.

xpdavid commented on src/main/java/teammates/ui/webapi/action/GetFeedbackSessionsAction.java:
```diff
@@ -20,12 +23,27 @@ protected AuthType getMinAuthLevel() {
+
+        if (isRequestAsStudent) {
+            if (courseId == null) {
+                gateKeeper.verifyStudentPrivileges(logic.getAccount(userInfo.getId()));
```
I think the correct way is to assert userInfo.isStudent to be true.

If you take a look at `verifyStudentPrivileges`, it actually do `verifyLoggedInUserPrivileges` which is not correct.

xpdavid commented on src/main/java/teammates/ui/webapi/action/GetFeedbackSessionsAction.java:
```diff
@@ -20,12 +23,27 @@ protected AuthType getMinAuthLevel() {
-            CourseAttributes courseAttributes = logic.getCourse(courseId);
-            gateKeeper.verifyAccessible(logic.getInstructorForGoogleId(courseId, userInfo.getId()), courseAttributes);
+            if (courseId == null) {
+                gateKeeper.verifyInstructorPrivileges(logic.getAccount(userInfo.getId()));
```
Can just assert isInstructor is true? (and save one DB read `logic.getAccount`)?

xpdavid commented on src/main/java/teammates/ui/webapi/action/GetFeedbackSessionsAction.java:
```diff
@@ -34,15 +52,26 @@ public ActionResult execute() {
 
-            if (isInRecycleBin) {
-                feedbackSessionAttributes = logic.getSoftDeletedFeedbackSessionsListForInstructors(instructors);
+            if (isRequestAsStudent) {
```
There are some fields in FeedbackSessionAttribute that students are not supposed to see. See `GetFeedbackSessionAction` to get a idea about what field to hide.

xpdavid commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +32,249 @@ protected String getRequestMethod() {
+        InstructorAttributes instructor2OfCourse1 = typicalBundle.instructors.get("instructor2OfCourse1");
+        loginAsInstructor(instructor2OfCourse1.googleId);
+
+        // The presence of IS_IN_RECYCLE_BIN flag is just used to indicate perform this request as instructor.
```
This is a sign that there are some problems with the design.

Previously I thought we could distinguish the request from instructor/student by just `IS_IN_RECYCLE_BIN`. It seems that it is not true anymore.

Let's see if we can use entityType (use GetCoursesAction as a reference)?

xpdavid replied:
So you should remove the comment?

junming403 commented:
Refactored hide attributes, make use of entityType. Ready for review

xpdavid commented on src/main/java/teammates/ui/webapi/output/FeedbackSessionData.java:
```diff
@@ -197,4 +197,18 @@ public long getCreatedAtTimestamp() {
     }
+
+    /**
+     * hide some attributes for student entity.
```
```suggestion
     * hide some attributes to student.
```

xpdavid commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +32,280 @@ protected String getRequestMethod() {
+
+        assertEquals(5, fsData.getFeedbackSessions().size());
+        for (FeedbackSessionData sessionData : fsData.getFeedbackSessions()) {
+            assertEquals(instructor2OfCourse1.getCourseId(), sessionData.getCourseId());
```
You should check that the information is hidden?

xpdavid commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +32,280 @@ protected String getRequestMethod() {
+
+        assertEquals(2, fsData.getFeedbackSessions().size());
+        for (FeedbackSessionData sessionData : fsData.getFeedbackSessions()) {
+            assertEquals(student2InCourse2.getCourse(), sessionData.getCourseId());
```
You should have at least one test which verify the full attributes returned by the API endpoint.

xpdavid commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +32,280 @@ protected String getRequestMethod() {
+        loginAsInstructor(instructor1OfCourse2.googleId);
+
+        String[] instructorParam = {
+                Const.ParamsNames.IS_IN_RECYCLE_BIN, "false",
```
The parameter can be removed?

junming403 commented:
@xpdavid  Ready for review, now all tests verified the content of feedback sessions, and for students, we checked some information is hidden and verified the rest

xpdavid commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +37,369 @@ protected String getRequestMethod() {
+    private void assertAllInstructorSessionsMatch(FeedbackSessionsData sessionsData,
+                                                  List<FeedbackSessionAttributes> expectedSessions) {
+
+        for (FeedbackSessionData sessionData : sessionsData.getFeedbackSessions()) {
```
You should check that the size of them are the same?

xpdavid commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +37,369 @@ protected String getRequestMethod() {
+    private void assertAllStudentSessionsMatch(FeedbackSessionsData sessionsData,
+                                               List<FeedbackSessionAttributes> expectedSessions) {
+
+        for (FeedbackSessionData sessionData : sessionsData.getFeedbackSessions()) {
```
Same here?

xpdavid commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +37,369 @@ protected String getRequestMethod() {
+
+        assertEquals(1, fsData.getFeedbackSessions().size());
+        FeedbackSessionData fs = fsData.getFeedbackSessions().get(0);
+        assertNotNull(fs.getDeletedAtTimestamp());
```
This assertion is no necessary as it is tested in `assertAllInformationMatch `

xpdavid commented on src/main/java/teammates/ui/webapi/action/GetFeedbackSessionsAction.java:
```diff
@@ -20,35 +24,72 @@ protected AuthType getMinAuthLevel() {
     @Override
     public ActionResult execute() {
         String courseId = getRequestParamValue(Const.ParamsNames.COURSE_ID);
+        String entityType = getNonNullRequestParamValue(Const.ParamsNames.ENTITY_TYPE);
```
There is at least one place in the frontend that use the API. I think you need to add entity type as a parameter for them. It would be better to abstract the method in service to reduce code duplication.

xpdavid commented on src/test/java/teammates/test/cases/webapi/GetFeedbackSessionsActionTest.java:
```diff
@@ -20,16 +37,370 @@ protected String getRequestMethod() {
+    }
+
+    @Test
+    protected void testExecute_noEntityType_shouldFail() {
```
```suggestion
    protected void testExecute_unknownEntityType_shouldFail() {
```

xpdavid commented on src/web/services/feedback-sessions.service.ts:
```diff
@@ -57,4 +57,29 @@ export class FeedbackSessionsService {
+  /**
+   * Gets all sessions with given entity type by calling API.
+   */
+  getFeedbackSessions(entityType: string, isInRecycleBin?: string, courseId?: string): Observable<FeedbackSessions> {
```
How about separating the methods?
```
getFeedbackSessionsForStudent()
getFeedbackSessionsForInstructor()
getFeedbackSessionsInRecycleBinForInstructor()
```

xpdavid commented on src/main/java/teammates/ui/webapi/output/FeedbackSessionData.java:
```diff
@@ -197,4 +197,18 @@ public long getCreatedAtTimestamp() {
+    /**
+     * hide some attributes to student.
+     */
+    public void hideInformationForStudent() {
```
The `GetFeedbackSessionAction` can also be changed to use the method?

xpdavid commented on src/main/java/teammates/ui/webapi/output/FeedbackSessionData.java:
```diff
@@ -197,4 +197,18 @@ public long getCreatedAtTimestamp() {
     }
+
+    /**
+     * hide some attributes to student.
```
```suggestion
     * Hides some attributes to student.
```

xpdavid commented on src/main/java/teammates/ui/webapi/action/GetFeedbackSessionsAction.java:
```diff
@@ -20,35 +26,77 @@ protected AuthType getMinAuthLevel() {
 
-        return new JsonResult(new FeedbackSessionsData(feedbackSessionAttributes));
+        if (entityType.equals(Const.EntityType.STUDENT)) {
+            feedbackSessionAttributes = feedbackSessionAttributes.stream().filter(session ->
```
Try `.isVisible()` and add an comment `// hide session not visible to student`?

junming403 commented:
Sorry for the careless mistake, all updated now! @xpdavid
